### PR TITLE
lint, fmt: support `tags` for approaches and concepts

### PR DESCRIPTION
**Pull request by ErikSchierboom**
- Lint tags for approaches
- Lint tags for concepts
- Don't check tags for articles
- Try fmt tags

---

For background, see the [forum thread][1].

[1]: https://forum.exercism.org/t/proposal-adding-tag-detection-to-analyzers/6931/7

### CODE REVIEW

File: src/fmt/approaches.nim [hunk 1]
```diff
@@ -35,6 +49,10 @@ func addApproach(result: var string; val: ApproachConfig; indentLevel = 1) =
   result.addArray("authors", val.authors, indentLevel + 1)
   if val.contributors.len > 0:
     result.addArray("contributors", val.contributors, indentLevel + 1)
+  if val.tags.isSome():
+    let tags = val.tags.get()
+    if tags.all.len + tags.`any`.len + tags.`not`.len > 0:
+      result.addApproachTags(tags, indentLevel + 1)
```
Comment (ErikSchierboom) [inline src/fmt/approaches.nim:55]: Let me know if you know of a more elegant way to do this:

File: src/lint/approaches_and_articles.nim [hunk 2]
```diff
@@ -41,6 +41,7 @@ proc isValidApproachOrArticle(data: JsonNode, context: string,
       hasArrayOfStrings(data, "authors", path, context, uniqueValues = true),
       hasArrayOfStrings(data, "contributors", path, context,
                         isRequired = false),
+      if checkTags: hasValidTags(data, path) else: true,
```
Comment (ErikSchierboom) [inline src/lint/approaches_and_articles.nim]: We currently only support tags for approaches, and not for articles

File: src/types_approaches_config.nim [hunk 3]
```diff
@@ -1,3 +1,5 @@
+import std/options
+import pkg/jsony
```
Comment (ee7) [inline src/types_approaches_config.nim]:
```suggestion
```

Now exported from helpers, which is imported in this file.

File: src/types_track_config.nim [hunk 4]
```diff
@@ -78,10 +78,16 @@ type
     indentSize*: int
     highlightjsLanguage*: string
 
+  ConceptTags* = object
+    all*: seq[string]
+    `any`*: seq[string]
+    `not`*: seq[string]
```
Comment (ee7) [inline src/types_track_config.nim:84]: The fields here are the same as `ApproachConfigTags`. Should we have a single source of truth?

File: src/lint/approaches_and_articles.nim [hunk 5]
```diff
@@ -63,6 +64,12 @@ proc isValidApproachOrArticle(data: JsonNode, context: string,
                   "there is no corresponding snippet file at the below location"
         result.setFalseIfFileMissingOrEmpty(snippetPath, msg)
 
+proc isValidApproach(data: JsonNode, context: string, path: Path): bool =
+  result = isValidApproachOrArticle(data, context, path, checkTags = true)
```
Comment (ee7) [inline src/lint/approaches_and_articles.nim]:
```suggestion
  isValidApproachOrArticle(data, context, path, checkTags = true)
```

Let's avoid `result =` for clear one-liners.

File: src/lint/approaches_and_articles.nim [hunk 6]
```diff
@@ -63,6 +64,12 @@ proc isValidApproachOrArticle(data: JsonNode, context: string,
...
+
+proc isValidArticle(data: JsonNode, context: string, path: Path): bool =
+  result = isValidApproachOrArticle(data, context, path, checkTags = false)
```
Comment (ee7) [inline src/lint/approaches_and_articles.nim]:
```suggestion
  isValidApproachOrArticle(data, context, path, checkTags = false)
```

File: src/types_track_config.nim (hunk 4 above)
Comment (ErikSchierboom) [reply src/types_track_config.nim:84]: We could. What would we call that file?

File: src/fmt/approaches.nim (hunk 1 above)
Comment (ee7) [reply src/fmt/approaches.nim:55]: We could iterate over the fields at compile time and generate something equivalent. Something like this (untested): 

```nim
for field in tags.fields:
  if field.len > 0:
    result.addApproachTags(tags, indentLevel + 1)
    break
```

But we didn't do that anywhere else. So let's not do it here. I hope for all this custom serialization mess to be handled much more elegantly in the future.

File: src/types_track_config.nim (hunk 4 above)
Comment (ErikSchierboom) [reply src/types_track_config.nim:84]: Feel free to extract it to a separate file.

Comment (ee7) [reply src/types_track_config.nim:84]: >  file

I meant having a single (renamed) object for both.

Maybe this?:

- remove the `ConceptTags` definition
- rename remaining `ConceptTags` instances to `AnalyzerTags`
- rename all `ApproachConfigTags` instances to `AnalyzerTags`?

 Or whatever you think of this as conceptually.

Comment (ee7) [reply src/types_track_config.nim:84]: Oh, I see. The definitions were in different files.

Comment (ee7) [reply src/types_track_config.nim:84]: Well, I think the options are:

1. The status quo
2. Add a separate `types_common.nim` 
3. Make `types_track_config` contain:
 
    ```nim
    from types_approaches_config import AnalyzerTags
    ```

Let's just stick with option 1 for now?

Comment (ErikSchierboom) [reply src/types_track_config.nim:84]: I think option 2 is fine. But we can do that as a follow-up 🤷

Comment (ErikSchierboom) [reply src/types_track_config.nim:84]: Let's do it as a follow-up. There might be more types that we can re-use.

File: src/lint/validators.nim [hunk 7]
```diff
@@ -568,6 +568,30 @@ proc hasInteger*(data: JsonNode; key: string; path: Path; context = "";
...
+        var anyAllLen = 0
+
+        if data[k].hasKey("all"):
+          anyAllLen += data[k]["all"].len
+
+        if data[k].hasKey("any"):
+          anyAllLen += data[k]["any"].len
+
+        if anyAllLen == 0:
+          result.setFalseAndPrint("There must be at least one element in the `all` or `any` fields in the `tags` object", path)
```
Comment (ee7) [inline src/lint/validators.nim]:
```suggestion
          const msg = "There must be at least one element in the `all` " &
                      "or `any` fields in the `tags` object"
          result.setFalseAndPrint(msg, path)
```

**Review by ee7 — APPROVED** [review]
No diff to output of `configlet lint` on any track.

Diff in output of `configlet fmt` on exactly one track:

#### csharp
```diff
-The below paths are relative to '/home/foo/exercism-tracks/csharp'
-Not formatted: config.json
-Not formatted: exercises/practice/bank-account/.approaches/config.json
-Not formatted: exercises/practice/bob/.approaches/config.json
-Not formatted: exercises/practice/collatz-conjecture/.approaches/config.json
-Not formatted: exercises/practice/darts/.approaches/config.json
-Not formatted: exercises/practice/difference-of-squares/.approaches/config.json
-Not formatted: exercises/practice/diffie-hellman/.approaches/config.json
-Not formatted: exercises/practice/gigasecond/.approaches/config.json
-Not formatted: exercises/practice/grade-school/.approaches/config.json
-Not formatted: exercises/practice/grains/.approaches/config.json
-Not formatted: exercises/practice/hamming/.approaches/config.json
-Not formatted: exercises/practice/isogram/.approaches/config.json
-Not formatted: exercises/practice/leap/.approaches/config.json
-Not formatted: exercises/practice/pangram/.approaches/config.json
-Not formatted: exercises/practice/parallel-letter-frequency/.approaches/config.json
-Not formatted: exercises/practice/perfect-numbers/.approaches/config.json
-Not formatted: exercises/practice/phone-number/.approaches/config.json
-Not formatted: exercises/practice/pig-latin/.approaches/config.json
-Not formatted: exercises/practice/protein-translation/.approaches/config.json
-Not formatted: exercises/practice/proverb/.approaches/config.json
-Not formatted: exercises/practice/raindrops/.approaches/config.json
-Not formatted: exercises/practice/reverse-string/.approaches/config.json
-Not formatted: exercises/practice/rna-transcription/.approaches/config.json
-Not formatted: exercises/practice/series/.approaches/config.json
-Not formatted: exercises/practice/sieve/.approaches/config.json
-Not formatted: exercises/practice/two-fer/.approaches/config.json
-Not formatted: exercises/practice/yacht/.approaches/config.json
+Every exercise config is already formatted!

```

File: src/lint/track_config.nim [hunk 8]
```diff
@@ -135,6 +135,7 @@ proc isValidConcept(data: JsonNode; context: string; path: Path): bool =
       hasString(data, "uuid", path, context, checkIsUuid = true),
       hasString(data, "slug", path, context, maxLen = 255, checkIsKebab = true),
       hasString(data, "name", path, context, maxLen = 255),
+      hasValidTags(data, path),
```
Comment (ee7) [inline src/lint/track_config.nim]: There's a blocking bug here. There is already a proc in this file named `hasValidTags`, which lints the track-level `tags` (like `typing/static`). It also has the same signature!

```console
$ git grep --break --heading 'proc hasValidTags'
src/lint/track_config.nim
268:proc hasValidTags(data: JsonNode; path: Path): bool =

src/lint/validators.nim
571:proc hasValidTags*(data: JsonNode; path: Path): bool =
```

Could you rename the new one in `validators.nim`? `hasValidAnalyzerTags` is fine by me, if that's accurate terminology.

Comment (ee7) [reply src/lint/track_config.nim]: Nim silently uses the in-file definition:

```console
$ cat a.nim
import b

proc foo =
  echo "a"

foo()
$ cat b.nim
proc foo* =
  echo "b"
$ nim r --hints:off a.nim
/tmp/a.nim(1, 8) Warning: imported and not used: 'b' [UnusedImport]
a
```

This is like variable shadowing, which Nim considers an idiom, but I admit it's much more unfortunate here.

So I believe the the new tags were being linted with the wrong validator.

Comment (ErikSchierboom) [reply src/lint/track_config.nim]: Good catch!

**Review by ee7 — APPROVED** [review]

**Pull request merged**